# The hemolymph of *Biomphalaria* snail vectors of schistosomiasis supports a diverse microbiome

## Abstract
The microbiome – the microorganism community that is found on or within an organism’s body – is increasingly recognized to shape many aspects of its host biology and is a key determinant of health and disease. Microbiomes modulate the capacity of insect disease vectors (mosquitos, tsetse flies, sandflies) to transmit parasites and disease. We investigate the diversity and abundance of microorganisms within the hemolymph (i.e. blood) of *Biomphalaria* snails, the intermediate host for *Schistosoma mansoni*, using Illumina MiSeq sequencing of the V4 variable region of the bacterial 16S ribosomal DNA. We sampled hemolymph from 5 snails from 6 different laboratory populations of *B. glabrata* and one population of *B. alexandrina*. We observed 281.34 ± 0.81 amplicon sequence variants (ASVs) per snail, where ASVs are defined as rDNA sequences differing by >3%. There were significant differences in microbiome composition at the level of individual snails, snail populations and snail species. Snail microbiomes were dominated by bacteria from the phyla Proteobacteria or Bacteroidetes while water microbiomes from snail tanks were dominated by bacteria from the phylum Actinobacteria. We investigated the absolute bacterial load using qPCR: hemolymph samples contained between ~200 to ~60,000 16S rDNA.µL<sup>-1</sup> independently of the snail populations or species, equivalent to 2,620 ± 321 bacteria per µL. We speculate that the microbiome may represent a critical, but unexplored intermediary in the snail-schistosome interaction as hemolymph is in very close contact to the parasite at each step of its development.

Full article is available [here](https://doi.org/XX)

## Code

The code used to analyze the data presented in the mansucript is available in this repository.

## Prerequisites

Two dependencies must be first installed before running the Jupyter notebook:
* A [conda](https://docs.conda.io/en/latest/) distribution, like [miniconda](https://docs.conda.io/en/latest/miniconda.html),
* [Jupyter Notebook](https://jupyter.readthedocs.io/en/latest/install.html) with a [bash kernel](https://github.com/takluyver/bash_kernel).

Once this is done, run the first cell of the Jupyter notebook from this repository to install a dedicated conda environment and related R packages.
