{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Snail hemolymph microbiome\n",
    "\n",
    "\n",
    "## Aim\n",
    "\n",
    "This notebook contains the different steps to install the environment, download the data and perform the data analysis. \n",
    "\n",
    "\n",
    "Several folders present or will be created during the analysis. Here is the list of the folders and their content:\n",
    "* **data**: Raw data files used for the analysis.\n",
    "* **.env**: Files needed to create appropriate environment.\n",
    "* **graphs**: Graphical representation of the data. If not existing, this will be created during the analysis\n",
    "* **results**: Files that are generating through data processing. If not existing, this will be created during the analysis\n",
    "* **scripts**: Scripts used for the analysis.\n",
    "\n",
    "\n",
    "## Environment and data\n",
    "\n",
    "### Creating environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check if conda available\n",
    "[[ ! $(which conda 2> /dev/null) ]] && echo \"conda not available in \\$PATH. Exiting...\" && exit 1\n",
    "\n",
    "# Creating conda environment\n",
    "conda env create -f .env/env.yml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This cell must be run each time a new session of Jupyter is run."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(ubiome_hml) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Activate the environment\n",
    "source $(sed \"s,/bin/conda,,\" <<<$CONDA_EXE)/etc/profile.d/conda.sh\n",
    "conda activate ubiome_hml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(ubiome_hml) Bioconductor version 3.8 (BiocManager 1.30.9), ?BiocManager::install for help\n",
      "\n",
      "Attaching package: ‘BiocManager’\n",
      "\n",
      "The following object is masked from ‘package:devtools’:\n",
      "\n",
      "    install\n",
      "\n",
      "Skipping install of 'microbiome' from a bioc_git2r remote, the SHA1 (11ae6af6) has not changed since last install.\n",
      "  Use `force = TRUE` to force installation\n",
      "Bioconductor version 3.8 (BiocManager 1.30.9), R 3.5.1 (2018-07-02)\n",
      "Installing package(s) 'phyloseq'\n",
      "trying URL 'https://bioconductor.org/packages/3.8/bioc/src/contrib/phyloseq_1.26.1.tar.gz'\n",
      "Content type 'application/x-gzip' length 5494354 bytes (5.2 MB)\n",
      "==================================================\n",
      "downloaded 5.2 MB\n",
      "\n",
      "* installing *source* package ‘phyloseq’ ...\n",
      "** R\n",
      "** data\n",
      "** inst\n",
      "** byte-compile and prepare package for lazy loading\n",
      "** help\n",
      "*** installing help indices\n",
      "** building package indices\n",
      "** installing vignettes\n",
      "** testing if installed package can be loaded\n",
      "* DONE (phyloseq)\n",
      "\n",
      "The downloaded source packages are in\n",
      "\t‘/tmp/RtmpFx6F5J/downloaded_packages’\n",
      "Updating HTML index of packages in '.Library'\n",
      "Making 'packages.html' ... done\n",
      "Downloading package from url: https://cran.revolutionanalytics.com/src/contrib/phylogram_2.1.0.tar.gz\n",
      "* installing *source* package ‘phylogram’ ...\n",
      "** package ‘phylogram’ successfully unpacked and MD5 sums checked\n",
      "** R\n",
      "** inst\n",
      "** byte-compile and prepare package for lazy loading\n",
      "** help\n",
      "*** installing help indices\n",
      "** building package indices\n",
      "** installing vignettes\n",
      "** testing if installed package can be loaded\n",
      "* DONE (phylogram)\n",
      "Skipping install of 'pairwiseAdonis' from a github remote, the SHA1 (6e097139) has not changed since last install.\n",
      "  Use `force = TRUE` to force installation\n",
      "Skipping install of 'taxa' from a github remote, the SHA1 (49969dcb) has not changed since last install.\n",
      "  Use `force = TRUE` to force installation\n",
      "Skipping install of 'qiime2R' from a github remote, the SHA1 (08e3b305) has not changed since last install.\n",
      "  Use `force = TRUE` to force installation\n",
      "Downloading Bioconductor repo https://git.bioconductor.org/packages/ALDEx2\n",
      "\u001b[32m✔\u001b[39m  \u001b[38;5;247mchecking for file ‘/tmp/RtmpFx6F5J/file53a533f29a9a/DESCRIPTION’\u001b[39m\u001b[36m\u001b[36m (399ms)\u001b[36m\u001b[39m\n",
      "\u001b[38;5;247m─\u001b[39m\u001b[38;5;247m  \u001b[39m\u001b[38;5;247mpreparing ‘ALDEx2’:\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[32m✔\u001b[39m  \u001b[38;5;247mchecking DESCRIPTION meta-information\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[38;5;247m─\u001b[39m\u001b[38;5;247m  \u001b[39m\u001b[38;5;247mchecking for LF line-endings in source and make files and shell scripts\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[38;5;247m─\u001b[39m\u001b[38;5;247m  \u001b[39m\u001b[38;5;247mchecking for empty or unneeded directories\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[38;5;247m─\u001b[39m\u001b[38;5;247m  \u001b[39m\u001b[38;5;247mlooking to see if a ‘data/datalist’ file should be added\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[38;5;247m─\u001b[39m\u001b[38;5;247m  \u001b[39m\u001b[38;5;247mbuilding ‘ALDEx2_1.16.0.tar.gz’\u001b[39m\u001b[36m\u001b[39m\n",
      "   \n",
      "* installing *source* package ‘ALDEx2’ ...\n",
      "** R\n",
      "** data\n",
      "** inst\n",
      "** byte-compile and prepare package for lazy loading\n",
      "** help\n",
      "*** installing help indices\n",
      "** building package indices\n",
      "** installing vignettes\n",
      "** testing if installed package can be loaded\n",
      "* DONE (ALDEx2)\n",
      "Skipping install of 'CoDaSeq' from a github remote, the SHA1 (73f77cef) has not changed since last install.\n",
      "  Use `force = TRUE` to force installation\n",
      "(ubiome_hml) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Installing needed R packages\n",
    "Rscript \".env/R package dependencies.R\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Downloading sequencing data\n",
    "\n",
    "This step downloads the fastq files of the different samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data directory\n",
    "ldir=\"data/libraries\"\n",
    "[[ ! -d \"$ldir\" ]] && mkdir -p \"$ldir\"\n",
    "\n",
    "# Bioproject\n",
    "bioproject=PRJNA613098\n",
    "\n",
    "# Download related information to data project\n",
    "wget -q -O runinfo \"http://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?save=efetch&rettype=runinfo&db=sra&term=${bioproject}\"\n",
    "\n",
    "# Field of interest (library name and weblink)\n",
    "fdn=$(head -n 1 runinfo | tr \",\" \"\\n\" | grep -w -n \"LibraryName\" | cut -d \":\" -f 1)\n",
    "fdr=$(head -n 1 runinfo | tr \",\" \"\\n\" | grep -w -n \"Run\" | cut -d \":\" -f 1)\n",
    "\n",
    "# Download fastq files\n",
    "while read line\n",
    "do\n",
    "    # Filename and download link\n",
    "    fln=$(cut -d \",\" -f $fdn <<<$line)\n",
    "    run=$(cut -d \",\" -f $fdr <<<$line)\n",
    "    \n",
    "    # Download\n",
    "    echo \"$fln\"\n",
    "    #wget -P \"$ldir\" -O \"$fln\" \"http://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?cmd=dload&run_list=${run}&format=fastq\"\n",
    "    fastq-dump -O \"$ldir\" --split-files \"$run\"\n",
    "    \n",
    "    mv \"$ldir/${run}_1.fastq\" \"$ldir/${fln}_R1.fastq\"\n",
    "    mv \"$ldir/${run}_2.fastq\" \"$ldir/${fln}_R2.fastq\"\n",
    "        \n",
    "done < <(tail -n +2 runinfo | head -1)\n",
    "\n",
    "# Compress files\n",
    "pigz \"$ldir/\"*\n",
    "\n",
    "rm runinfo"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Downloading database\n",
    "\n",
    "The Silva database is used to assign taxonomy to the ASVs generated from the sequencing data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(ubiome_hml) (ubiome_hml) (ubiome_hml) (ubiome_hml) (ubiome_hml) (ubiome_hml) --2020-04-10 12:41:44--  https://www.arb-silva.de/fileadmin/silva_databases/qiime/Silva_132_release.zip\n",
      "Resolving www.arb-silva.de... 194.95.6.29\n",
      "Connecting to www.arb-silva.de|194.95.6.29|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 2466176198 (2.3G) [application/zip]\n",
      "Saving to: ‘database/Silva_132_release.zip’\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The notebook server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--NotebookApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "NotebookApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "2020-04-10 15:10:56 (269 KB/s) - ‘database/Silva_132_release.zip’ saved [2466176198/2466176198]\n",
      "\n",
      "(ubiome_hml) Archive:  database/Silva_132_release.zip\n",
      "   creating: database/SILVA_132_QIIME_release/\n",
      "   creating: database/SILVA_132_QIIME_release/core_alignment/\n",
      "  inflating: database/SILVA_132_QIIME_release/core_alignment/80_core_alignment.fna  \n",
      "   creating: database/SILVA_132_QIIME_release/raw_data/\n",
      "  inflating: database/SILVA_132_QIIME_release/raw_data/.DS_Store  \n",
      "   creating: database/__MACOSX/\n",
      "   creating: database/__MACOSX/SILVA_132_QIIME_release/\n",
      "   creating: database/__MACOSX/SILVA_132_QIIME_release/raw_data/\n",
      "  inflating: database/__MACOSX/SILVA_132_QIIME_release/raw_data/._.DS_Store  \n",
      "  inflating: database/SILVA_132_QIIME_release/raw_data/raw_taxonomy_split_by_level.txt.zip  \n",
      "  inflating: database/SILVA_132_QIIME_release/raw_data/raw_taxonomy.txt.zip  \n",
      "  inflating: database/SILVA_132_QIIME_release/raw_data/initial_reads_SILVA132.fna.zip  \n",
      "  inflating: database/SILVA_132_QIIME_release/raw_data/gap_positions_SILVA132.txt  \n",
      "   creating: database/SILVA_132_QIIME_release/rep_set/\n",
      "   creating: database/SILVA_132_QIIME_release/rep_set/rep_set_16S_only/\n",
      "   creating: database/SILVA_132_QIIME_release/rep_set/rep_set_16S_only/90/\n",
      "  inflating: database/SILVA_132_QIIME_release/rep_set/rep_set_16S_only/90/silva_132_90_16S.fna  \n",
      "   creating: database/__MACOSX/SILVA_132_QIIME_release/rep_set/\n",
      "   creating: database/__MACOSX/SILVA_132_QIIME_release/rep_set/rep_set_16S_only/\n",
      "  inflating: database/__MACOSX/SILVA_132_QIIME_release/rep_set/rep_set_16S_only/._90  \n",
      "  inflating: database/SILVA_132_QIIME_release/rep_set/rep_set_16S_only/.DS_Store  \n",
      "  inflating: database/__MACOSX/SILVA_132_QIIME_release/rep_set/rep_set_16S_only/._.DS_Store  \n",
      "   creating: database/SILVA_132_QIIME_release/rep_set/rep_set_16S_only/94/\n",
      "  inflating: database/SILVA_132_QIIME_release/rep_set/rep_set_16S_only/94/silva_132_94_16S.fna  \n",
      "   creating: database/SILVA_132_QIIME_release/rep_set/rep_set_16S_only/97/\n",
      "  inflating: database/SILVA_132_QIIME_release/rep_set/rep_set_16S_only/97/silva_132_97_16S.fna  \n",
      "   creating: database/SILVA_132_QIIME_release/rep_set/rep_set_16S_only/99/\n",
      "  inflating: database/SILVA_132_QIIME_release/rep_set/rep_set_16S_only/99/silva_132_99_16S.fna  \n",
      "  inflating: database/__MACOSX/SILVA_132_QIIME_release/rep_set/._rep_set_16S_only  \n",
      "   creating: database/SILVA_132_QIIME_release/rep_set/rep_set_18S_only/\n",
      "  inflating: database/SILVA_132_QIIME_release/rep_set/rep_set_18S_only/.DS_Store  \n",
      "   creating: database/__MACOSX/SILVA_132_QIIME_release/rep_set/rep_set_18S_only/\n",
      "  inflating: database/__MACOSX/SILVA_132_QIIME_release/rep_set/rep_set_18S_only/._.DS_Store  \n",
      "   creating: database/SILVA_132_QIIME_release/rep_set/rep_set_18S_only/94/\n",
      "  inflating: database/SILVA_132_QIIME_release/rep_set/rep_set_18S_only/94/silva_132_94_18S.fna  \n",
      "   creating: database/SILVA_132_QIIME_release/rep_set/rep_set_18S_only/97/\n",
      "  inflating: database/SILVA_132_QIIME_release/rep_set/rep_set_18S_only/97/silva_132_97_18S.fna  \n",
      "   creating: database/SILVA_132_QIIME_release/rep_set/rep_set_18S_only/99/\n",
      "  inflating: database/SILVA_132_QIIME_release/rep_set/rep_set_18S_only/99/silva_132_99_18S.fna  \n",
      "   creating: database/SILVA_132_QIIME_release/rep_set/rep_set_18S_only/90/\n",
      "  inflating: database/SILVA_132_QIIME_release/rep_set/rep_set_18S_only/90/silva_132_90_18S.fna  \n",
      "  inflating: database/__MACOSX/SILVA_132_QIIME_release/rep_set/rep_set_18S_only/._90  \n",
      "   creating: database/SILVA_132_QIIME_release/rep_set/rep_set_all/\n",
      "   creating: database/SILVA_132_QIIME_release/rep_set/rep_set_all/90/\n",
      "  inflating: database/SILVA_132_QIIME_release/rep_set/rep_set_all/90/silva132_90.fna  \n",
      "   creating: database/__MACOSX/SILVA_132_QIIME_release/rep_set/rep_set_all/\n",
      "  inflating: database/__MACOSX/SILVA_132_QIIME_release/rep_set/rep_set_all/._90  \n",
      "   creating: database/SILVA_132_QIIME_release/rep_set/rep_set_all/94/\n",
      "  inflating: database/SILVA_132_QIIME_release/rep_set/rep_set_all/94/silva132_94.fna  \n",
      "   creating: database/SILVA_132_QIIME_release/rep_set/rep_set_all/97/\n",
      "  inflating: database/SILVA_132_QIIME_release/rep_set/rep_set_all/97/silva132_97.fna  \n",
      "   creating: database/SILVA_132_QIIME_release/rep_set/rep_set_all/99/\n",
      "  inflating: database/SILVA_132_QIIME_release/rep_set/rep_set_all/99/silva132_99.fna  \n",
      "  inflating: database/SILVA_132_QIIME_release/rep_set/rep_set_all/.DS_Store  \n",
      "  inflating: database/__MACOSX/SILVA_132_QIIME_release/rep_set/rep_set_all/._.DS_Store  \n",
      "   creating: database/SILVA_132_QIIME_release/rep_set_aligned/\n",
      "   creating: database/SILVA_132_QIIME_release/rep_set_aligned/90/\n",
      "  inflating: database/SILVA_132_QIIME_release/rep_set_aligned/90/.DS_Store  \n",
      "   creating: database/__MACOSX/SILVA_132_QIIME_release/rep_set_aligned/\n",
      "   creating: database/__MACOSX/SILVA_132_QIIME_release/rep_set_aligned/90/\n",
      "  inflating: database/__MACOSX/SILVA_132_QIIME_release/rep_set_aligned/90/._.DS_Store  \n",
      "  inflating: database/SILVA_132_QIIME_release/rep_set_aligned/90/90_alignment.fna.zip  \n",
      "   creating: database/SILVA_132_QIIME_release/rep_set_aligned/94/\n",
      "  inflating: database/SILVA_132_QIIME_release/rep_set_aligned/94/94_alignment.fna.zip  \n",
      "   creating: database/SILVA_132_QIIME_release/rep_set_aligned/97/\n",
      "  inflating: database/SILVA_132_QIIME_release/rep_set_aligned/97/97_alignment.fna.zip  \n",
      "   creating: database/SILVA_132_QIIME_release/rep_set_aligned/99/\n",
      "  inflating: database/SILVA_132_QIIME_release/rep_set_aligned/99/99_alignment.fna.zip  \n",
      "   creating: database/SILVA_132_QIIME_release/taxonomy/\n",
      "   creating: database/SILVA_132_QIIME_release/taxonomy/16S_only/\n",
      "   creating: database/SILVA_132_QIIME_release/taxonomy/16S_only/99/\n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/16S_only/99/raw_taxonomy.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/16S_only/99/taxonomy_7_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/16S_only/99/taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/16S_only/99/majority_taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/16S_only/99/majority_taxonomy_7_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/16S_only/99/consensus_taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/16S_only/99/consensus_taxonomy_7_levels.txt  \n",
      "   creating: database/SILVA_132_QIIME_release/taxonomy/16S_only/90/\n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/16S_only/90/raw_taxonomy.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/16S_only/90/taxonomy_7_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/16S_only/90/taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/16S_only/90/majority_taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/16S_only/90/majority_taxonomy_7_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/16S_only/90/consensus_taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/16S_only/90/consensus_taxonomy_7_levels.txt  \n",
      "   creating: database/__MACOSX/SILVA_132_QIIME_release/taxonomy/\n",
      "   creating: database/__MACOSX/SILVA_132_QIIME_release/taxonomy/16S_only/\n",
      "  inflating: database/__MACOSX/SILVA_132_QIIME_release/taxonomy/16S_only/._90  \n",
      "   creating: database/SILVA_132_QIIME_release/taxonomy/16S_only/94/\n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/16S_only/94/raw_taxonomy.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/16S_only/94/taxonomy_7_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/16S_only/94/taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/16S_only/94/majority_taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/16S_only/94/majority_taxonomy_7_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/16S_only/94/consensus_taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/16S_only/94/consensus_taxonomy_7_levels.txt  \n",
      "   creating: database/SILVA_132_QIIME_release/taxonomy/16S_only/97/\n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/16S_only/97/raw_taxonomy.txt  \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/16S_only/97/taxonomy_7_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/16S_only/97/taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/16S_only/97/majority_taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/16S_only/97/majority_taxonomy_7_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/16S_only/97/consensus_taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/16S_only/97/consensus_taxonomy_7_levels.txt  \n",
      "  inflating: database/__MACOSX/SILVA_132_QIIME_release/taxonomy/._16S_only  \n",
      "   creating: database/SILVA_132_QIIME_release/taxonomy/18S_only/\n",
      "   creating: database/SILVA_132_QIIME_release/taxonomy/18S_only/90/\n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/18S_only/90/raw_taxonomy.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/18S_only/90/taxonomy_7_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/18S_only/90/taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/18S_only/90/majority_taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/18S_only/90/majority_taxonomy_7_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/18S_only/90/consensus_taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/18S_only/90/consensus_taxonomy_7_levels.txt  \n",
      "   creating: database/__MACOSX/SILVA_132_QIIME_release/taxonomy/18S_only/\n",
      "  inflating: database/__MACOSX/SILVA_132_QIIME_release/taxonomy/18S_only/._90  \n",
      "   creating: database/SILVA_132_QIIME_release/taxonomy/18S_only/94/\n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/18S_only/94/raw_taxonomy.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/18S_only/94/taxonomy_7_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/18S_only/94/taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/18S_only/94/majority_taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/18S_only/94/majority_taxonomy_7_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/18S_only/94/consensus_taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/18S_only/94/consensus_taxonomy_7_levels.txt  \n",
      "   creating: database/SILVA_132_QIIME_release/taxonomy/18S_only/97/\n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/18S_only/97/raw_taxonomy.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/18S_only/97/taxonomy_7_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/18S_only/97/taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/18S_only/97/majority_taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/18S_only/97/majority_taxonomy_7_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/18S_only/97/consensus_taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/18S_only/97/consensus_taxonomy_7_levels.txt  \n",
      "   creating: database/SILVA_132_QIIME_release/taxonomy/18S_only/99/\n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/18S_only/99/raw_taxonomy.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/18S_only/99/taxonomy_7_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/18S_only/99/taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/18S_only/99/majority_taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/18S_only/99/majority_taxonomy_7_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/18S_only/99/consensus_taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/18S_only/99/consensus_taxonomy_7_levels.txt  \n",
      "   creating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/\n",
      "   creating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/90/\n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/90/raw_taxonomy.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/90/taxonomy_7_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/90/taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/90/consensus_taxonomy_7_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/90/consensus_taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/90/majority_taxonomy_7_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/90/majority_taxonomy_all_levels.txt  \n",
      "   creating: database/__MACOSX/SILVA_132_QIIME_release/taxonomy/taxonomy_all/\n",
      "  inflating: database/__MACOSX/SILVA_132_QIIME_release/taxonomy/taxonomy_all/._90  \n",
      "   creating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/94/\n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/94/raw_taxonomy.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/94/taxonomy_7_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/94/taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/94/consensus_taxonomy_7_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/94/consensus_taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/94/majority_taxonomy_7_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/94/majority_taxonomy_all_levels.txt  \n",
      "   creating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/97/\n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/97/raw_taxonomy.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/97/taxonomy_7_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/97/taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/97/consensus_taxonomy_7_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/97/consensus_taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/97/majority_taxonomy_7_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/97/majority_taxonomy_all_levels.txt  \n",
      "   creating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/99/\n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/99/raw_taxonomy.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/99/taxonomy_7_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/99/taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/99/consensus_taxonomy_7_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/99/consensus_taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/99/majority_taxonomy_7_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/99/majority_taxonomy_all_levels.txt  \n",
      "  inflating: database/SILVA_132_QIIME_release/taxonomy/taxonomy_all/.DS_Store  \n",
      "  inflating: database/__MACOSX/SILVA_132_QIIME_release/taxonomy/taxonomy_all/._.DS_Store  \n",
      "   creating: database/SILVA_132_QIIME_release/trees/\n",
      "   creating: database/SILVA_132_QIIME_release/trees/90/\n",
      "  inflating: database/SILVA_132_QIIME_release/trees/90/90_otus.tre  \n",
      "   creating: database/__MACOSX/SILVA_132_QIIME_release/trees/\n",
      "  inflating: database/__MACOSX/SILVA_132_QIIME_release/trees/._90  \n",
      "   creating: database/SILVA_132_QIIME_release/trees/94/\n",
      "  inflating: database/SILVA_132_QIIME_release/trees/94/94_otus.tre  \n",
      "   creating: database/__MACOSX/SILVA_132_QIIME_release/trees/94/\n",
      "  inflating: database/__MACOSX/SILVA_132_QIIME_release/trees/94/._94_otus.tre  \n",
      "   creating: database/SILVA_132_QIIME_release/trees/97/\n",
      "  inflating: database/SILVA_132_QIIME_release/trees/97/97_otus.tre  \n",
      "   creating: database/__MACOSX/SILVA_132_QIIME_release/trees/97/\n",
      "  inflating: database/__MACOSX/SILVA_132_QIIME_release/trees/97/._97_otus.tre  \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   creating: database/SILVA_132_QIIME_release/trees/99/\n",
      "  inflating: database/SILVA_132_QIIME_release/trees/99/99_otus.tre  \n",
      "  inflating: database/SILVA_132_QIIME_release/Silva_132_notes.txt  \n",
      "  inflating: database/__MACOSX/SILVA_132_QIIME_release/._Silva_132_notes.txt  \n",
      "(ubiome_hml) (ubiome_hml) (ubiome_hml) \u001b[32mImported database/SILVA_132_QIIME_release/rep_set/rep_set_16S_only/99/silva_132_99_16S.fna as DNASequencesDirectoryFormat to database/silva_132_99_16S.qza\u001b[0m\n",
      "(ubiome_hml) (ubiome_hml) (ubiome_hml) \u001b[32mImported database/SILVA_132_QIIME_release/taxonomy/16S_only/99/taxonomy_all_levels.txt as HeaderlessTSVTaxonomyFormat to database/silva_132_99_16S_taxa.qza\u001b[0m\n",
      "(ubiome_hml) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Database directory\n",
    "dbdir=\"data/Silva db\"\n",
    "[[ ! -d \"$dbdir\" ]] && mkdir -p \"$dbdir\"\n",
    "\n",
    "\n",
    "# Download and extract the relevant Silva file\n",
    "wget -P \"$dbdir\" 'https://www.arb-silva.de/fileadmin/silva_databases/qiime/Silva_132_release.zip'\n",
    "unzip \"$dbdir/Silva_132_release.zip\" -d \"$dbdir\" && rm \"$dbdir/Silva_132_release.zip\"\n",
    "\n",
    "# Import the sequence database in Qiime format\n",
    "qiime tools import \\\n",
    "  --input-path \"$dbdir/SILVA_132_QIIME_release/rep_set/rep_set_16S_only/99/silva_132_99_16S.fna\" \\\n",
    "  --output-path \"$dbdir/silva_132_99_16S.qza\" \\\n",
    "  --type 'FeatureData[Sequence]'\n",
    "\n",
    "# Import the taxonomy database in Qiime format\n",
    "qiime tools import \\\n",
    "  --input-path \"$dbdir/SILVA_132_QIIME_release/taxonomy/16S_only/99/taxonomy_all_levels.txt\" \\\n",
    "  --output-path \"$dbdir/silva_132_99_16S_taxa.qza\" \\\n",
    "  --type 'FeatureData[Taxonomy]' \\\n",
    "  --input-format HeaderlessTSVTaxonomyFormat"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Qiime pipeline\n",
    "\n",
    "This section process the data to generate ASVs and assign taxonomy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(ubiome_hml) (ubiome_hml) (ubiome_hml) (ubiome_hml) (ubiome_hml) (ubiome_hml) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Qiime output directory\n",
    "qdir=\"results/1-qiime\"\n",
    "[[ ! -d \"$qdir\" ]] && mkdir -p \"$qdir\"\n",
    "\n",
    "# Metadata file\n",
    "metadata=\"data/sample-metadata.tsv\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sequencing data\n",
    "\n",
    "If sequencing data are present, they will be imported as a Qiime artifact."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(ubiome_hml) (ubiome_hml) (ubiome_hml) (ubiome_hml) (ubiome_hml) (ubiome_hml) (ubiome_hml) (ubiome_hml) (ubiome_hml) (ubiome_hml) (ubiome_hml) (ubiome_hml) \u001b[32mImported results/1-qiime/manifest as PairedEndFastqManifestPhred33V2 to results/1-qiime/demux-paired-end.qza\u001b[0m\n",
      "(ubiome_hml) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Check for sequencing data\n",
    "[[ ! $(find \"$ldir\" -type f -name *fastq.gz) ]] && echo  \"No sequencing data. Exiting...\" && exit 1\n",
    "\n",
    "# Create the manifest for importing data in artefact\n",
    "## source: https://docs.qiime2.org/2019.4/tutorials/importing/#fastq-manifest-formats\n",
    "for i in $(ls \"$ldir\"/* | sed \"s,_R[12].fastq.*,,g\" | uniq)\n",
    "do\n",
    "    nm=$(sed \"s,$ldir/,,\" <<<$i)\n",
    "    fl=$(ls -1 \"$PWD/$i\"* | tr \"\\n\" \"\\t\")\n",
    "\n",
    "    echo -e \"$nm\\t$fl\"\n",
    "done > \"$qdir/manifest\"\n",
    "\n",
    "# Add header\n",
    "sed -i \"1s/^/sample-id\\tforward-absolute-filepath\\treverse-absolute-filepath\\n/\" \"$qdir/manifest\"\n",
    "\n",
    "# Import data\n",
    "## source: https://docs.qiime2.org/2019.4/tutorials/importing/\n",
    "qiime tools import \\\n",
    "  --type 'SampleData[PairedEndSequencesWithQuality]' \\\n",
    "  --input-path \"$qdir/manifest\" \\\n",
    "  --input-format PairedEndFastqManifestPhred33V2 \\\n",
    "  --output-path \"$qdir/demux-paired-end.qza\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note**:\n",
    "* No need to remove adapters and barcodes. This has been done during `bcl2fastq`. This can be checked using `grep`.\n",
    "* Importing with `--input-format PairedEndFastqManifestPhred33V2` instead of `--input-format CasavaOneEightSingleLanePerSampleDirFmt` for [custom sample names](https://docs.qiime2.org/2019.4/tutorials/importing/#fastq-manifest-formats).\n",
    "\n",
    "### Data quality\n",
    "\n",
    "To assess data quality, we need to generate a visualization to check data quality. The visualization can be view on [Qiime2 website](https://view.qiime2.org/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(ubiome_hml) \u001b[32mSaved Visualization to: results/1-qiime/demux-paired-end.qzv\u001b[0m\n",
      "(ubiome_hml) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Make a summary to check read quality\n",
    "qiime demux summarize \\\n",
    "  --i-data \"$qdir/demux-paired-end.qza\" \\\n",
    "  --o-visualization \"$qdir/demux-paired-end.qzv\""
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# This cell must be run in a terminal in place of viewing the visualization on the Qiime2 website.\n",
    "[[ -z $DISPLAY ]] && echo \"You need a X11 connection.\"\n",
    "qdir=\"results/1-qiime\"\n",
    "qiime tools view \"$qdir/demux-paired-end.qzv\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note**: read quality drops toward the end but are still above 10. So no trimming done.\n",
    "\n",
    "### Sequence clustering and denoising\n",
    "\n",
    "This steps generates ASVs from the sequencing data. This step is perform by the `dada2` module."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(ubiome_hml) (ubiome_hml) (ubiome_hml) \u001b[32mSaved FeatureTable[Frequency] to: results/1-qiime/table.qza\u001b[0m\n",
      "\u001b[32mSaved FeatureData[Sequence] to: results/1-qiime/rep-seqs.qza\u001b[0m\n",
      "\u001b[32mSaved SampleData[DADA2Stats] to: results/1-qiime/denoising-stats.qza\u001b[0m\n",
      "(ubiome_hml) (ubiome_hml) (ubiome_hml) \u001b[32mSaved Visualization to: results/1-qiime/denoising-stats.qzv\u001b[0m\n",
      "(ubiome_hml) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "[[ -f \"$qdir/table.qza\" ]] && echo \"ASV file (table.qza) exists already. Exiting...\" && exit 1\n",
    "\n",
    "# Run dada2\n",
    "qiime dada2 denoise-paired \\\n",
    "  --i-demultiplexed-seqs \"$qdir/demux-paired-end.qza\" \\\n",
    "  --p-trunc-len-f 250 \\\n",
    "  --p-trunc-len-r 250 \\\n",
    "  --p-max-ee 5 \\\n",
    "  --p-n-threads 0 \\\n",
    "  --o-table \"$qdir/table.qza\" \\\n",
    "  --o-representative-sequences \"$qdir/rep-seqs.qza\" \\\n",
    "  --o-denoising-stats \"$qdir/denoising-stats.qza\"\n",
    "\n",
    "# Add metadata information to the denoising stats\n",
    "qiime metadata tabulate \\\n",
    "    --m-input-file \"$metadata\" \\\n",
    "    --m-input-file \"$qdir/denoising-stats.qza\" \\\n",
    "    --o-visualization \"$qdir/denoising-stats.qzv\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note:** Visual inspection shows that optimal number of features is 25079 which allows retaining 2,633,295 (46.33%) features in 105 (99.06%) samples. This leads to the exclusion of one sample: `Ba.Water.1`.\n",
    "\n",
    "### Taxonomy identification\n",
    "\n",
    "This step assigns taxonomy to the ASVs generated."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureData[Taxonomy] to: results/1-qiime/rep-seqs_taxa.qza\u001b[0m\n",
      "(ubiome_hml) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "qiime feature-classifier classify-consensus-vsearch \\\n",
    "  --i-query \"$qdir/rep-seqs.qza\" \\\n",
    "  --i-reference-reads \"$dbdir/silva_132_99_16S.qza\" \\\n",
    "  --i-reference-taxonomy \"$dbdir/silva_132_99_16S_taxa.qza\" \\\n",
    "  --p-perc-identity 0.97 \\\n",
    "  --p-threads $(nproc) \\\n",
    "  --o-classification \"$qdir/rep-seqs_taxa.qza\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Phylogeny\n",
    "\n",
    "This step generates a phylogeny from the ASVs ([source](https://chmi-sops.github.io/mydoc_qiime2.html))."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(ubiome_hml) (ubiome_hml) (ubiome_hml) \u001b[32mSaved FeatureData[AlignedSequence] to: results/1-qiime/aligned-rep-seqs.qza\u001b[0m\n",
      "(ubiome_hml) (ubiome_hml) (ubiome_hml) \u001b[32mSaved FeatureData[AlignedSequence] to: results/1-qiime/masked-aligned-rep-seqs.qza\u001b[0m\n",
      "(ubiome_hml) (ubiome_hml) (ubiome_hml) \u001b[32mSaved Phylogeny[Unrooted] to: results/1-qiime/unrooted-tree.qza\u001b[0m\n",
      "(ubiome_hml) (ubiome_hml) (ubiome_hml) \u001b[32mSaved Phylogeny[Rooted] to: results/1-qiime/rooted-tree.qza\u001b[0m\n",
      "(ubiome_hml) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "[[ -f \"$qdir/rooted-tree.qza\" ]] && echo \"A tree file (rooted-tree.qza) exists already. Exiting...\" && exit 1\n",
    "\n",
    "# Multiple seqeunce alignment using Mafft\n",
    "qiime alignment mafft \\\n",
    "    --i-sequences \"$qdir/rep-seqs.qza\" \\\n",
    "    --o-alignment \"$qdir/aligned-rep-seqs.qza\"\n",
    "\n",
    "# Masking (or filtering) the alignment to remove positions that are highly variable. These positions are generally considered to add noise to a resulting phylogenetic tree.\n",
    "qiime alignment mask \\\n",
    "    --i-alignment \"$qdir/aligned-rep-seqs.qza\" \\\n",
    "    --o-masked-alignment \"$qdir/masked-aligned-rep-seqs.qza\"\n",
    "\n",
    "# Creating tree using the Fasttree program\n",
    "qiime phylogeny fasttree \\\n",
    "    --i-alignment \"$qdir/masked-aligned-rep-seqs.qza\" \\\n",
    "    --o-tree \"$qdir/unrooted-tree.qza\"\n",
    "\n",
    "# Root the tree using the longest root\n",
    "qiime phylogeny midpoint-root \\\n",
    "    --i-tree \"$qdir/unrooted-tree.qza\" \\\n",
    "    --o-rooted-tree \"$qdir/rooted-tree.qza\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Library analysis\n",
    "\n",
    "Analyze Qiime visualization and generate table summarizing number of reads after each step of the pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Output:  ../results/1-qiime/lib-seq-stat (supp. table 1).tsv \n",
      "Output:  ../graphs/seq.ct_1.pdf \n",
      "Output:  ../graphs/seq.final_1.pdf \n",
      "Output:  ../graphs/seq.ct_2.pdf \n",
      "Output:  ../graphs/seq.final_2.pdf \n",
      "Output:  ../results/1-qiime/lib-stat.tsv \n",
      "(ubiome_hml) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "scripts/library_stats.R"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Microbiome diversity\n",
    "\n",
    "This step analyze ASV data using different methods (rarefaction curve, $\\alpha$ and $\\beta$ diversity). Technical replicates are also evaluated to show the robustness of the library generation method. Details about the methods used are in the R script. Analysis of the results are in the manuscript."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading packages...\n",
      "Loading functions...\n",
      "Setting variables...\n",
      "Warning messages:\n",
      "1: Expected 7 pieces. Additional pieces discarded in 1045 rows [3, 4, 5, 7, 9, 12, 16, 23, 27, 28, 33, 34, 35, 42, 43, 48, 49, 50, 51, 55, ...]. \n",
      "2: Expected 7 pieces. Missing pieces filled with `NA` in 1643 rows [1, 2, 6, 8, 10, 11, 13, 14, 15, 17, 18, 19, 20, 21, 22, 24, 25, 26, 29, 30, ...]. \n",
      "rarefying sample Ba.1.1\n",
      "rarefying sample Ba.1.2\n",
      "rarefying sample Ba.2.1\n",
      "rarefying sample Ba.2.2\n",
      "rarefying sample Ba.3.1\n",
      "rarefying sample Ba.3.2\n",
      "rarefying sample Ba.4.1\n",
      "rarefying sample Ba.4.2\n",
      "rarefying sample Ba.5.1\n",
      "rarefying sample Ba.5.2\n",
      "rarefying sample Ba.Water.1\n",
      "rarefying sample Ba.Water.2\n",
      "rarefying sample Bg121.1.1\n",
      "rarefying sample Bg121.1.2\n",
      "rarefying sample Bg121.2.1\n",
      "rarefying sample Bg121.2.2\n",
      "rarefying sample Bg121.3.1\n",
      "rarefying sample Bg121.3.2\n",
      "rarefying sample Bg121.4.1\n",
      "rarefying sample Bg121.4.2\n",
      "rarefying sample Bg121.5.1\n",
      "rarefying sample Bg121.5.2\n",
      "rarefying sample Bg121.Water.1\n",
      "rarefying sample Bg121.Water.2\n",
      "rarefying sample Bg26.1.1\n",
      "rarefying sample Bg26.1.2\n",
      "rarefying sample Bg26.2.1\n",
      "rarefying sample Bg26.2.2\n",
      "rarefying sample Bg26.3.1\n",
      "rarefying sample Bg26.3.2\n",
      "rarefying sample Bg26.4.1\n",
      "rarefying sample Bg26.4.2\n",
      "rarefying sample Bg26.5.1\n",
      "rarefying sample Bg26.5.2\n",
      "rarefying sample Bg26.Water.1\n",
      "rarefying sample Bg26.Water.2\n",
      "rarefying sample Bg36.1.1\n",
      "rarefying sample Bg36.1.2\n",
      "rarefying sample Bg36.2.1\n",
      "rarefying sample Bg36.2.2\n",
      "rarefying sample Bg36.3.1\n",
      "rarefying sample Bg36.3.2\n",
      "rarefying sample Bg36.4.1\n",
      "rarefying sample Bg36.4.2\n",
      "rarefying sample Bg36.5.1\n",
      "rarefying sample Bg36.5.2\n",
      "rarefying sample Bg36.Water.1\n",
      "rarefying sample Bg36.Water.2\n",
      "rarefying sample BgBRE.1.1\n",
      "rarefying sample BgBRE.1.2\n",
      "rarefying sample BgBRE.2.1\n",
      "rarefying sample BgBRE.2.2\n",
      "rarefying sample BgBRE.3.1\n",
      "rarefying sample BgBRE.3.2\n",
      "rarefying sample BgBRE.4.1\n",
      "rarefying sample BgBRE.4.2\n",
      "rarefying sample BgBRE.5.1\n",
      "rarefying sample BgBRE.5.2\n",
      "rarefying sample BgBRE.Water.1\n",
      "rarefying sample BgBRE.Water.2\n",
      "rarefying sample BgBS90.1.1\n",
      "rarefying sample BgBS90.1.2\n",
      "rarefying sample BgBS90.2.1\n",
      "rarefying sample BgBS90.2.2\n",
      "rarefying sample BgBS90.3.1\n",
      "rarefying sample BgBS90.3.2\n",
      "rarefying sample BgBS90.4.1\n",
      "rarefying sample BgBS90.4.2\n",
      "rarefying sample BgBS90.5.1\n",
      "rarefying sample BgBS90.5.2\n",
      "rarefying sample BgBS90.Water.1\n",
      "rarefying sample BgBS90.Water.2\n",
      "rarefying sample BgNMRI.1.1\n",
      "rarefying sample BgNMRI.1.2\n",
      "rarefying sample BgNMRI.2.1\n",
      "rarefying sample BgNMRI.2.2\n",
      "rarefying sample BgNMRI.3.1\n",
      "rarefying sample BgNMRI.3.2\n",
      "rarefying sample BgNMRI.4.1\n",
      "rarefying sample BgNMRI.4.2\n",
      "rarefying sample BgNMRI.5.1\n",
      "rarefying sample BgNMRI.5.2\n",
      "rarefying sample BgNMRI.Water.1\n",
      "rarefying sample BgNMRI.Water.2\n",
      "Observed richness\n",
      "Other forms of richness\n",
      "Diversity\n",
      "Evenness\n",
      "Dominance\n",
      "Rarity\n",
      "Observed number of ASVs in snails:\n",
      "\t- mean:  281.3429 \n",
      "\t- standard error:  0.8064399 \n",
      "\n",
      "Observed number of ASVs in water:\n",
      "\t- mean:  145.5714 \n",
      "\t- standard error:  4.037862 \n",
      "\n",
      "There were 28 warnings (use warnings() to see them)\n",
      "`set.seed(26017)` was used to initialize repeatable random subsampling.\n",
      "Please record this for your records so others can reproduce.\n",
      "Try `set.seed(26017); .Random.seed` for the full vector\n",
      "...\n",
      "850OTUs were removed because they are no longer \n",
      "present in any sample after random subsampling\n",
      "\n",
      "...\n",
      "`set.seed(26017)` was used to initialize repeatable random subsampling.\n",
      "Please record this for your records so others can reproduce.\n",
      "Try `set.seed(26017); .Random.seed` for the full vector\n",
      "...\n",
      "634OTUs were removed because they are no longer \n",
      "present in any sample after random subsampling\n",
      "\n",
      "...\n",
      "Warning messages:\n",
      "1: In matrix(tree$edge[order(tree$edge[, 1]), ][, 2], byrow = TRUE,  :\n",
      "  data length [4093] is not a sub-multiple or multiple of the number of rows [2047]\n",
      "2: In matrix(tree$edge[order(tree$edge[, 1]), ][, 2], byrow = TRUE,  :\n",
      "  data length [4093] is not a sub-multiple or multiple of the number of rows [2047]\n",
      "`set.seed(26017)` was used to initialize repeatable random subsampling.\n",
      "Please record this for your records so others can reproduce.\n",
      "Try `set.seed(26017); .Random.seed` for the full vector\n",
      "...\n",
      "850OTUs were removed because they are no longer \n",
      "present in any sample after random subsampling\n",
      "\n",
      "...\n",
      "`set.seed(26017)` was used to initialize repeatable random subsampling.\n",
      "Please record this for your records so others can reproduce.\n",
      "Try `set.seed(26017); .Random.seed` for the full vector\n",
      "...\n",
      "634OTUs were removed because they are no longer \n",
      "present in any sample after random subsampling\n",
      "\n",
      "...\n",
      "`set.seed(26017)` was used to initialize repeatable random subsampling.\n",
      "Please record this for your records so others can reproduce.\n",
      "Try `set.seed(26017); .Random.seed` for the full vector\n",
      "...\n",
      "850OTUs were removed because they are no longer \n",
      "present in any sample after random subsampling\n",
      "\n",
      "...\n",
      "Scale for 'fill' is already present. Adding another scale for 'fill', which\n",
      "will replace the existing scale.\n",
      "Scale for 'fill' is already present. Adding another scale for 'fill', which\n",
      "will replace the existing scale.\n",
      "Scale for 'fill' is already present. Adding another scale for 'fill', which\n",
      "will replace the existing scale.\n",
      "Scale for 'fill' is already present. Adding another scale for 'fill', which\n",
      "will replace the existing scale.\n",
      "`set.seed(26017)` was used to initialize repeatable random subsampling.\n",
      "Please record this for your records so others can reproduce.\n",
      "Try `set.seed(26017); .Random.seed` for the full vector\n",
      "...\n",
      "634OTUs were removed because they are no longer \n",
      "present in any sample after random subsampling\n",
      "\n",
      "...\n",
      "Scale for 'fill' is already present. Adding another scale for 'fill', which\n",
      "will replace the existing scale.\n",
      "Scale for 'fill' is already present. Adding another scale for 'fill', which\n",
      "will replace the existing scale.\n",
      "Scale for 'fill' is already present. Adding another scale for 'fill', which\n",
      "will replace the existing scale.\n",
      "Scale for 'fill' is already present. Adding another scale for 'fill', which\n",
      "will replace the existing scale.\n",
      "Warning messages:\n",
      "1: Removed 8 rows containing missing values (geom_text). \n",
      "2: Removed 8 rows containing missing values (geom_text). \n",
      "3: Removed 8 rows containing missing values (geom_text). \n",
      "4: Removed 8 rows containing missing values (geom_text). \n",
      "5: Removed 8 rows containing missing values (geom_text). \n",
      "Creating a temporary cluster...done:\n",
      "socket cluster with 15 nodes on host ‘localhost’\n",
      "Multiscale bootstrap... Done.\n",
      "Scale for 'x' is already present. Adding another scale for 'x', which will\n",
      "replace the existing scale.\n",
      "Scale for 'x' is already present. Adding another scale for 'x', which will\n",
      "replace the existing scale.\n",
      "Unassigned bacteria proportion in snails:\n",
      "\t- mean:  0.04187556 \n",
      "\t- range (min max):  0.0008122229 0.1632689 \n",
      "\n",
      "Unassigned bacteria proportion in water:\n",
      "\t- mean:  0.01105316 \n",
      "\t- range (min max):  0 0.1017087 \n",
      "\n",
      "Warning messages:\n",
      "1: In psmelt(physeq) : The sample variables: \n",
      "Species\n",
      " have been renamed to: \n",
      "sample_Species\n",
      "to avoid conflicts with taxonomic rank names.\n",
      "2: In psmelt(physeq) : The sample variables: \n",
      "Species\n",
      " have been renamed to: \n",
      "sample_Species\n",
      "to avoid conflicts with taxonomic rank names.\n",
      "Warning message:\n",
      "Transformation introduced infinite values in discrete y-axis \n",
      "Warning message:\n",
      "Transformation introduced infinite values in discrete y-axis \n",
      "(ubiome_hml) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "Rscript scripts/microbiome_diversity.R"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Functional inference\n",
    "\n",
    "Analysis following this [tutorial](https://github.com/picrust/picrust2/wiki/PICRUSt2-Tutorial-(v2.1.4-beta)#pathway-level-inference)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(ubiome_hml) (ubiome_hml) (ubiome_hml) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# PiCRUST output directory\n",
    "pdir=\"results/2-picrust2\"\n",
    "[[ ! -d \"$pdir\" ]] && mkdir -p \"$pdir\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(ubiome_hml) \u001b[32mSaved FeatureTable[Frequency] to: results/1-qiime/table_rep1.qza\u001b[0m\n",
      "\u001b[32mSaved FeatureTable[Frequency] to: results/1-qiime/table_rep2.qza\u001b[0m\n",
      "(ubiome_hml) (ubiome_hml) (ubiome_hml) (ubiome_hml) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Filter table to retain sample from each replicate\n",
    "for i in {1..2}\n",
    "do\n",
    "    awk -v i=$i 'NR==1; $5 == i {print}' \"$metadata\" > \"$qdir/.metadata\"\n",
    "    \n",
    "    qiime feature-table filter-samples \\\n",
    "        --i-table \"$qdir/table.qza\"  \\\n",
    "        --m-metadata-file \"$qdir/.metadata\" \\\n",
    "        --o-filtered-table \"$qdir/table_rep$i.qza\"\n",
    "done\n",
    "\n",
    "# Clean\n",
    "rm \"$qdir/.metadata\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mExported results/1-qiime/rep-seqs.qza as DNASequencesDirectoryFormat to directory results/2-picrust2/\u001b[0m\n",
      "(ubiome_hml) (ubiome_hml) \u001b[32mExported results/1-qiime/table_rep1.qza as BIOMV210DirFmt to directory results/2-picrust2/rep1\u001b[0m\n",
      "\u001b[32mExported results/1-qiime/table_rep2.qza as BIOMV210DirFmt to directory results/2-picrust2/rep2\u001b[0m\n",
      "(ubiome_hml) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "qiime tools export \\\n",
    "    --input-path \"$qdir/rep-seqs.qza\" \\\n",
    "    --output-path \"$pdir/\"\n",
    "\n",
    "for i in {1..2}\n",
    "do\n",
    "    qiime tools export \\\n",
    "        --input-path \"$qdir/table_rep$i.qza\" \\\n",
    "        --output-path \"$pdir/rep$i\"\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(ubiome_hml) (ubiome_hml) (ubiome_hml) (ubiome_hml) (ubiome_hml) (ubiome_hml) (ubiome_hml) (ubiome_hml) 136\n",
      "(ubiome_hml) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Place reads into reference tree\n",
    "place_seqs.py -s \"$pdir/dna-sequences.fasta\" -o \"$pdir/out.tre\" -p $(nproc) \\\n",
    "    --intermediate \"$pdir/intermediate/place_seqs\"\n",
    "\n",
    "# Hidden-state prediction of gene families\n",
    "hsp.py -i 16S -t \"$pdir/out.tre\" -o \"$pdir/marker_predicted_and_nsti.tsv.gz\" -p $(nproc) -n\n",
    "hsp.py -i EC -t \"$pdir/out.tre\" -o \"$pdir/EC_predicted.tsv.gz\" -p $(nproc) -n\n",
    "\n",
    "# Number of outliers\n",
    "zcat \"$pdir/marker_predicted_and_nsti.tsv.gz\"  | tail -n +2 | awk '$3 >= 2' | wc -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(ubiome_hml) 136 of 2688 ASVs were above the max NSTI cut-off of 2.0 and were removed.\n",
      "136 of 2688 ASVs were above the max NSTI cut-off of 2.0 and were removed.\n",
      "136 of 2688 ASVs were above the max NSTI cut-off of 2.0 and were removed.\n",
      "136 of 2688 ASVs were above the max NSTI cut-off of 2.0 and were removed.\n",
      "(ubiome_hml) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Run for each replicate\n",
    "for i in {1..2}\n",
    "do\n",
    "    # Generate metagenome predictions\n",
    "    metagenome_pipeline.py \\\n",
    "        -i \"$pdir/rep$i/feature-table.biom\" \\\n",
    "        -m \"$pdir/marker_predicted_and_nsti.tsv.gz\" \\\n",
    "        -f \"$pdir/EC_predicted.tsv.gz\" \\\n",
    "        --max_nsti 2.0 \\\n",
    "        -o \"$pdir/rep$i/EC_metagenome_out\" \\\n",
    "        --strat_out --metagenome_contrib\n",
    "\n",
    "    # Pathway-level inference\n",
    "    pathway_pipeline.py \\\n",
    "        -i \"$pdir/rep$i/EC_metagenome_out/pred_metagenome_strat.tsv.gz\" \\\n",
    "        -o \"$pdir/rep$i/pathways_out\" -p $(nproc)\n",
    "\n",
    "    #Add functional descriptions\n",
    "    add_descriptions.py -i \"$pdir/rep$i/EC_metagenome_out/pred_metagenome_unstrat.tsv.gz\" -m EC \\\n",
    "                        -o \"$pdir/rep$i/EC_metagenome_out/pred_metagenome_unstrat_descrip.tsv.gz\"\n",
    "\n",
    "    add_descriptions.py -i \"$pdir/rep$i/pathways_out/path_abun_unstrat.tsv.gz\" -m METACYC \\\n",
    "                        -o \"$pdir/rep$i/pathways_out/path_abun_unstrat_descrip.tsv.gz\"\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading required package: zCompositions\n",
      "Loading required package: MASS\n",
      "Loading required package: NADA\n",
      "Loading required package: survival\n",
      "\n",
      "Attaching package: ‘NADA’\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    cor\n",
      "\n",
      "Loading required package: truncnorm\n",
      "Loading required package: car\n",
      "Loading required package: carData\n",
      "\n",
      "Attaching package: ‘igraph’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    union\n",
      "\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing zero removal\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing zero removal\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "aldex.clr: generating Monte-Carlo instances and clr values\n",
      "operating in serial mode\n",
      "removed rows with sums equal to zero\n",
      "computing center with all features\n",
      "data format is OK\n",
      "dirichlet samples complete\n",
      "transformation complete\n",
      "aldex.ttest: doing t-test\n",
      "running tests for each MC instance:\n",
      "|------------(25%)----------(50%)----------(75%)----------|\n",
      "aldex.effect: calculating effect sizes\n",
      "operating in serial mode\n",
      "sanity check complete\n",
      "rab.all  complete\n",
      "rab.win  complete\n",
      "rab of samples complete\n",
      "within sample difference calculated\n",
      "between group difference calculated\n",
      "group summaries calculated\n",
      "effect size calculated\n",
      "summarizing output\n",
      "(ubiome_hml) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "scripts/pathway_analysis.R"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Microbiome density\n",
    "\n",
    "Could the differences observed between population explained by microbe density?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(ubiome_hml) (ubiome_hml) (ubiome_hml) (ubiome_hml) (ubiome_hml) Raw 16S copy number\n",
      "\n",
      "16S copy number:\n",
      "\t- mean:  4490.034 \n",
      "\t- range (min max):  185.7344 59486.62 \n",
      "\n",
      "Data per population is not following normal distribution\n",
      "\n",
      "\tKruskal-Wallis rank sum test\n",
      "\n",
      "data:  qpcr.d[, 2] and qpcr.d[, 3]\n",
      "Kruskal-Wallis chi-squared = 26.494, df = 7, p-value = 0.0004108\n",
      "\n",
      "Pairwise comparisons of 16S copy number between population (mean of the replicates):  ab a ab a b ac c b \n",
      "\n",
      "\n",
      "Estimateed bacteria\n",
      "\n",
      "Bacteria number:\n",
      "\t- mean:  2405.149 \n",
      "\t- range (min max):  165.6501 16452.82 \n",
      "\n",
      "Data per population is not following normal distribution\n",
      "\n",
      "\tKruskal-Wallis rank sum test\n",
      "\n",
      "data:  asv.d[, 2] and asv.d[, 3]\n",
      "Kruskal-Wallis chi-squared = 18.294, df = 7, p-value = 0.01071\n",
      "\n",
      "Pairwise comparisons of estimated bacteria between population (mean of the replicates):  ab ab ab ab a ab b ab \n",
      "\n",
      "Mean bacteria number in snails:\n",
      "\t- mean:  2620.105 \n",
      "\t- standard error:  320.6233 \n",
      "\n",
      "Warning messages:\n",
      "1: Removed 6 rows containing non-finite values (stat_boxplot). \n",
      "2: Removed 6 rows containing non-finite values (stat_summary). \n",
      "3: Removed 2 rows containing non-finite values (stat_boxplot). \n",
      "4: Removed 2 rows containing non-finite values (stat_summary). \n",
      "(ubiome_hml) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# qPCR output directory\n",
    "ddir=\"data/qPCR\"\n",
    "[[ ! -d \"$ddir\" ]] && echo \"$ddir and qPCR data are missing. Exiting...\" && exit 1\n",
    "\n",
    "# Analyze data\n",
    "scripts/microbiome_density.R"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": false,
   "sideBar": true,
   "skip_h1_title": true,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "230px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
